Add bounded error Hausdorff distance (#5475)
CGAL/cgal: Polygon_mesh_processing/doc/Polygon_mesh_processing/Polygon_mesh_processing.txt

\snippet Poisson_surface_reconstruction_3/poisson_reconstruction_example.cpp PMP_distance_snippet

\subsection BoundedHD Bounded Hausdorff Distance

The function `CGAL::Polygon_mesh_processing::bounded_error_Hausdorff_distance()`
computes an estimate of the Hausdorff distance of two triangle meshes which is
bounded by a user-given error bound. Given two meshes tm1 and tm2, it follows

> the procedure given by \cgalCite{tang2009interactive}. Namely, an AABB tree is

Author: ```suggestion
the procedure given by \cgalCite{tang2009interactive}. Namely, bounded volume hierarchy (BVH) is
```

Reviewer: Done.

Reviewer 2: backtick the tms

built on tm1 and tm2 respectively. The tree on tm1 is used to iterate over all
triangles in tm1. Throughout the traversal, the procedure keeps track of a global
lower and upper bound on the Hausdorff distance respectively. For each triangle
t in tm1, by traversing the tree on tm2, it is estimated via the global bounds
whether t can still contribute to the actual Hausdorff distance. From this
process, a set of candidate triangles is selected.

The candidate triangles are subsequently subdivided and for each smaller
triangle, the tree on tm2 is traversed again. This is repeated until the
triangle is smaller than the user-given error bound, all vertices of the
triangle are projected onto the same triangle in tm2, or the triangle's upper
bound is lower than the global lower bound. After creation, the subdivided
triangles are added to the list of candidate triangles. Thereby, all candidate
triangles are processed until a triangle is found in which the Hausdorff
distance is realized or in which it is guaranteed to be realized within the

> user-given error bound.

Author: ```suggestion
user-given error bound.

In the current implementation, the BVH used in an AABB-tree and not the swept sphere volumes as used in the original implementation. This should explained the runtime difference observed with the original implementation.
```

Reviewer: Done.

\snippet Poisson_surface_reconstruction_3/poisson_reconstruction_example.cpp PMP_distance_snippet

\subsection BoundedHD Bounded Hausdorff Distance

The function `CGAL::Polygon_mesh_processing::bounded_error_Hausdorff_distance()`
computes an estimate of the Hausdorff distance of two triangle meshes which is
bounded by a user-given error bound. Given two meshes `tm1` and `tm2`, it follows
the procedure given by \cgalCite{tang2009interactive}. Namely, a bounded volume hierarchy (BVH) is

> built on `tm1` and `tm2` respectively. The tree on `tm1` is used to iterate over all

Author: ```suggestion
built on `tm1` and `tm2` respectively. The BVH on `tm1` is used to iterate over all
```

triangles in `tm1`. Throughout the traversal, the procedure keeps track of a global
lower and upper bound on the Hausdorff distance respectively. For each triangle

> `t` in `tm1`, by traversing the tree on `tm2`, it is estimated via the global bounds

Author: ```suggestion
`t` in `tm1`, by traversing the BVH on `tm2`, it is estimated via the global bounds
```

whether `t` can still contribute to the actual Hausdorff distance. From this
process, a set of candidate triangles is selected.

The candidate triangles are subsequently subdivided and for each smaller

> triangle, the tree on `tm2` is traversed again. This is repeated until the

Author: ```suggestion
triangle, the BVH on `tm2` is traversed again. This is repeated until the
```
